core/node/base_token_adjuster/src/base_token_l1_behaviour.rs

```diff
@@ -0,0 +1,331 @@
+                    }]
+                        .observe(start_time.elapsed());
+                    self.update_last_persisted_l1_ratio(
+                        BigDecimal::from(new_ratio.numerator.get())
+                            .div(BigDecimal::from(new_ratio.denominator.get())),
+                    );
+
+                    return Ok(());
+                }
+                Err(err) => {
+                    tracing::info!(
+                        "Failed to update base token multiplier on L1, attempt {}, base_fee_per_gas {}, priority_fee_per_gas {}: {}",
+                        attempt,
+                        base_fee_per_gas,
+                        priority_fee_per_gas,
+                        err
+                    );
+                    METRICS.l1_update_latency[&OperationResultLabels {
+                        result: OperationResult::Failure,
+                    }]
+                        .observe(start_time.elapsed());
+
+                    tokio::time::sleep(sleep_duration).await;
+                    prev_base_fee_per_gas = Some(base_fee_per_gas);
+                    prev_priority_fee_per_gas = Some(priority_fee_per_gas);
+                    last_error = Some(err)
+                }
+            }
+        }
+
+        let error_message = "Failed to update base token multiplier on L1";
+        Err(last_error
+            .map(|x| x.context(error_message))
+            .unwrap_or_else(|| anyhow::anyhow!(error_message)))
+    }
+
+    fn update_last_persisted_l1_ratio(&mut self, new_ratio: BigDecimal) {
+        match self {
+            BaseTokenL1Behaviour::UpdateOnL1 {
+                params: _,
+                ref mut last_persisted_l1_ratio,
+            } => *last_persisted_l1_ratio = Some(new_ratio),
+            BaseTokenL1Behaviour::NoOp => {}
+        };
+    }
+
+    async fn do_update_l1(
+        &self,
+        l1_params: &UpdateOnL1Params,
+        api_ratio: BaseTokenAPIRatio,
+        base_fee_per_gas: u64,
+        priority_fee_per_gas: u64,
+    ) -> anyhow::Result<Option<U256>> {
+        let fn_set_token_multiplier = l1_params
+            .chain_admin_contract
+            .function("setTokenMultiplier")
+            .context("`setTokenMultiplier` function must be present in the ChainAdmin contract")?;
+
+        let calldata = fn_set_token_multiplier
+            .encode_input(
+                &(
+                    Token::Address(l1_params.diamond_proxy_contract_address),
+                    Token::Uint(api_ratio.numerator.get().into()),
+                    Token::Uint(api_ratio.denominator.get().into()),
+                )
+                    .into_tokens(),
+            )
+            .context("failed encoding `setTokenMultiplier` input")?;
+
+        let nonce = (*l1_params.eth_client)
+            .as_ref()
+            .nonce_at_for_account(
+                l1_params.token_multiplier_setter_account_address,
+                BlockNumber::Pending,
+            )
```

Comment:
Maybe it's an edge case, but what if we send an underprice transaction, and then the server restarts? My understanding is that this transaction will be stuck forever. Do we want to think about this case?

Reply:
good catch, yes, let's use BlockNumber::Latest instead

Reply:
Well in the worst case multiplier is going to be updated twice then (which is not bad)

Reply:
The worst (and I'd say pretty realistic) case is that updating on L1 will be fully stuck

Reply:
IMO yes

Reply:
Ok just chatted to Roman - agreed, it should be Latest, changed and pushed. Thank you both for catching it.